### PR TITLE
[Dynamo] Graph break on functions using tensor out variants

### LINKED ISSUE #108021: AssertionError Triggered with Different Tensor Shapes of Input and Output in torch.cosh when Using torch.compile
### 🐛 Describe the bug

When using torch.compile on a model that employs torch.cosh with different tensor shapes for input and output, I encounter an AssertionError with the message expected size 2==1, stride 1==1 at dim=2.

🔄 To Reproduce
Here's a minimal example to reproduce the issue:
```python
import torch
from torch import nn

class CustomModel(nn.Module):
    def __init__(self):
        super(CustomModel, self).__init__()

    def forward(self, inputs):
        return torch.cosh(**inputs)

ip_size = [1]
input_tensor = torch.randn(ip_size)
out_tensor = torch.empty([2])
cuda_inputs = input_tensor.clone().to('cuda')
cuda_out = out_tensor.clone().to('cuda')

mymodel = CustomModel()
no_op_info= mymodel({'input': input_tensor, 'out': out_tensor})
mymodel.to('cuda')
op_info = torch.compile(mymodel)({'input': cuda_inputs, 'out': cuda_out})
print(op_info)
```
📃 Stack Trace
If applicable, please include the error stack trace for better context.
```
  File "/home/guihuan/.conda/envs/night/lib/python3.11/site-packages/torch/_inductor/codecache.py", line 401, in _run_from_cache
    return compiled_graph.compiled_artifact(inputs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/torchinductor_guihuan/ul/culusjx5t3qps4p5rwb3niikda3uwkfdfuscnrqvscqp3dbb4wtk.py", line 60, in call
    assert_size_stride(arg1_1, (1, ), (1, ))
AssertionError: expected size 2==1, stride 1==1 at dim=0
```
➕ Additional Context
If this behavior is intended, or if there are known bugs related to this issue, please feel free to close it. If additional information is needed, I'm more than willing to provide it.

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230826+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.11.0 (main, Mar  1 2023, 18:26:19) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-79-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2070
GPU 1: NVIDIA GeForce RTX 2070
GPU 2: NVIDIA GeForce RTX 2070
GPU 3: NVIDIA GeForce RTX 2070

Nvidia driver version: 535.86.10
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
CPU family:                      6
Model:                           63
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       2
Stepping:                        2
CPU max MHz:                     3200.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        4794.42
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
Virtualization:                  VT-x
L1d cache:                       512 KiB (16 instances)
L1i cache:                       512 KiB (16 instances)
L2 cache:                        4 MiB (16 instances)
L3 cache:                        40 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s):               1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+e6216047b8
[pip3] torch==2.1.0.dev20230826+cu118
[pip3] torchaudio==2.1.0.dev20230826+cu118
[pip3] torchvision==0.16.0.dev20230826+cu118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+e6216047b8          pypi_0    pypi
[conda] torch                     2.1.0.dev20230826+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230826+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230826+cu118          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #108021

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### CODE REVIEW

**Comment by yanboliang** [comment]
For the ```test_adamax``` failure, should we remove the out variant from optimizer code? @eellison

**Comment by eellison** [comment]
@yanboliang we could only refine the conditions to graph break so we only graph break if a resize is required, does that fix error ?

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -187,6 +187,9 @@ def load(self, tx):
     def erase(self):
         self._example = None
 
+    def __eq__(self, other):
```
Comment (eellison) [inline torch/_dynamo/variables/builder.py:190]: cant speak to whether this is valid or not

**Review by eellison — APPROVED** [review]
looks good ! thanks !

**Pull request closed without merging**